File: Java-Core/src/main/java/presto/type/TupleType.java
package presto.type;

import presto.error.SyntaxError;
import presto.runtime.Context;

public class TupleType extends CollectionType {

	static TupleType instance = new TupleType();
	
	public static TupleType instance() {
		return instance;
	}
	
	private TupleType() {
		super("Tuple", AnyType.instance());
	}

	@Override
	public Class<?> toJavaClass() {
		return null; // no equivalent
	}
	
	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return (other instanceof TupleType) || (other instanceof AnyType);
	}

	@Override
	public IType checkItem(Context context, IType other) throws SyntaxError {
		if(other==IntegerType.instance())
			return AnyType.instance();
		else
			return super.checkItem(context,other);
	}
	

	@Override
	public IType checkMember(Context context, String name) throws SyntaxError {
        if ("length".equals(name))
            return IntegerType.instance();
        else
            return super.checkMember(context, name);
}
	
	@Override
	public IType checkAdd(Context context, IType other, boolean tryReverse) throws SyntaxError {
		if(other instanceof TupleType || other instanceof ListType || other instanceof SetType)
			return this; 
		return super.checkAdd(context, other, tryReverse);
	}
	
	@Override
	public IType checkContains(Context context, IType other) throws SyntaxError {
		return BooleanType.instance(); 
	}
	
	@Override
	public IType checkContainsAllOrAny(Context context, IType other) throws SyntaxError {
		return BooleanType.instance(); 
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/type/TupleType.js
var NativeType = require("./NativeType").NativeType;
var BooleanType = require("./BooleanType").BooleanType;
var ListType = require("./ListType").ListType;
var IntegerType = require("./IntegerType").IntegerType;
var AnyType = require("./AnyType").AnyType;

function TupleType()  {
	NativeType.call(this, "Tuple");
	return this;
}

TupleType.prototype = Object.create(NativeType.prototype);
TupleType.prototype.constructor = TupleType;

TupleType.instance = new TupleType();

/*

@Override
public Class<?> toJavaClass() {
	return null; // no equivalent
}

@Override
public boolean isAssignableTo(Context context, IType other) {
	return (other instanceof TupleType) || (other instanceof AnyType);
}
*/
TupleType.prototype.checkItem = function(context, other) {
	if(other==IntegerType.instance) {
		return AnyType.instance;
	} else {
		return NativeType.prototype.checkItem.call(this, context, other);
	}
};

TupleType.prototype.checkMember = function(context, name) {
	if ("length"==name) {
		return IntegerType.instance;
	} else {
		return NativeType.prototype.checkMember.call(this, context, name);
	}
};

TupleType.prototype.checkAdd = function(context, other, tryReverse) {
	if(other instanceof TupleType || other instanceof ListType) {
		return this;
	} else {
		return NativeType.prototype.checkAdd.call(this, context, other, tryReverse);
	}
};

TupleType.prototype.checkContains = function(context, other) {
	return BooleanType.instance;
};

TupleType.prototype.checkContainsAllOrAny = function(context, other) {
	return BooleanType.instance;
};

exports.TupleType = TupleType;
